### DISCUSSION TITLE
Model loading taking too much time

### DISCUSSION THREAD

**Discussion by vibsid0986** [eclipse-langium/langium#1374] [Q&A]
Hello Team,

We conducted a trial by loading our model within a browser window containing a workspace comprising approximately 1000 files. The loading process consumed approximately 3 to 3.5 minutes. I am interested in understanding whether this aligns with the results observed during bulk testing on your end.

Please help us with it. 

Thanks in advance.

**Comment by msujew** [comment]
Hey @vibsid0986,

can you tell how large your model is in terms of bytesize/lines of code? A thousand files isn't really a lot in the context of Langium, but depending on the file size it might take a while. Our own benchmarks show that langium roughly processes 20000 LoC per second, but it heavily depends on the complexity of your grammar and the performance of your scoping algorithm, etc.

It might make sense to benchmark your workspace initialization to see where most of the time is spent. There are quite a few possible bottlenecks, some of which are language specific, others are runtime related. Giving a bit more context would help here.

**Reply by vibsid0986** [reply]
Hi @msujew ,

Thank you for your timely reply.

Below is the screenshot providing details on the average size of each file. Additionally, each file encompasses approximately 70 lines of code. For added context, our grammar structure is complex. Even with the absence of custom scoping/linking algorithms, the default implementations for it required a loading time of around 3 to 3.5 minutes. Kindly inform me if further information is necessary.

<img width="269" alt="image" src="https://github.com/eclipse-langium/langium/assets/98730468/4918cb6e-ff7d-45b9-b4d6-5b937e1bb221">

**Reply by msujew** [reply]
Ok, 3 minutes is definitely absurd for ~70.000 LoC. It should be done with a workspace of this size in a few seconds, even for more complicated grammars. Just for some context: We are using Langium in projects with 1mil+ LoC and it needs roughly a minute to initialize the workspace.

Can you provide a reproducible example for this issue? Or maybe a flamegraph of a benchmark? Without more context about your project, it's hard to recommend potential improvements for your performance issue.
